// @jatahworx/bhive-toolkits/ServiceDesigner/ndefault-all-nodes/nodes/server/MiddlewareWorkspaceDefaultNodes.js
let BaseComponent = require("@jatahworx/bhive-toolkits").BaseComponent;
class RouteMiddlewareStart extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_RouteMiddlewareStart", "RouteMiddlewareStart");
        this.viewType = 'server';
    }
}

class RouteMiddlewareEnd extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_RouteMiddlewareEnd", "RouteMiddlewareEnd");
        this.viewType = 'server';
    }
}

class Next extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_Next", "Next");
        this.viewType = 'server';
    }
}

class GlobalMiddlewareStart extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_DefaulGlobalMiddlewareStart", "GlobalMiddlewareStart");
        this.viewType = 'server';
    }
}

class GlobalMiddlewareEnd extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_GlobalMiddlewareEnd", "GlobalMiddlewareEnd");
        this.viewType = 'server';
    }
}

class GlobalNext extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_GlobalNext", "GlobalNext");
        this.viewType = 'server';
    }
}

module.exports = {
    RouteMiddlewareStart: RouteMiddlewareStart,
    RouteMiddlewareEnd: RouteMiddlewareEnd,
    Next: Next,
    GlobalMiddlewareStart: GlobalMiddlewareStart,
    GlobalMiddlewareEnd: GlobalMiddlewareEnd,
    GlobalNext: GlobalNext
};
