// @jatahworx/bhive-toolkits/ServiceDesigner/ndefault-mongodb/ndefault-mongodbOnStart.js
let BaseComponent = require('@jatahworx/bhive-toolkits').BaseComponent;

module.exports = class SQLOnStart extends BaseComponent {
    generateImports() {
        return [
          {
              library: './utils/ndefault-mongodb/Mongodb/MongoConnections',
              modules: [
                  'MongoConnections'
              ]
          }
        ];
      }
    
      generateSnippet() {
          return `
            let monogConnections = MongoConnections.getInstance();
            function isDBDisabled(flag) {
                return typeof flag === 'boolean' && flag;
            }
            const dbConfig = configNodes.default["db-config"];
            if (dbConfig) {
                const dbConfigsList = Object.keys(dbConfig);
                let ormConfig: any[] = [];
                for (let i = 0; i < dbConfigsList.length; i++) {
                    if (dbConfig[dbConfigsList[i]] && !isDBDisabled(dbConfig[dbConfigsList[i]].disabledb) && dbConfig[dbConfigsList[i]].dbOption) {
                        let dbOption = dbConfig[dbConfigsList[i]].dbOption;
                        if (dbOption.type === 'mongodb') {
                            await monogConnections.newConnection(dbOption, dbConfigsList[i]);
                        }
                    }
                }
            }
          `;
      }
    
};
