// @jatahworx/bhive-toolkits/angular-material/component/Table/TablePaginator.js
"use strict";
let Attribute = require("../../../core/Attributes");
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class TablePaginator extends AdvancedComponent {
  constructor() {
    const name = "ng-table-paginator";
    const designerTemplate = `<ng-table-paginator slot="paginator" block-copy><span class="component-placeholder">Table Paginator</span></ng-table-paginator>`;
    const paletteTemplate = "Table Paginator";
    super({
      name,
      designerTemplate,
      paletteTemplate,
      componentLabel: 'Paginator',
      templateUrl: 'articles/#!components-guide-7/tables',
      isAdvancedChild: false,
      visible: false
    });
    super.addAttribute(new Attribute({
      key: '[length]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '[pageSize]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '[pageSizeOptions]',
      value: '[5, 10, 25, 100]',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '(page)',
      value: '',
      type: 'kv'
    }));
    super.setType(AdvancedComponent.COMPONENT_TYPE_TITLES.NAVIGATION.val);
    super.composeTemplate({
      styles: `:host {
        position: relative;
        flex: 1 0 auto;
        min-width: 8em;
        min-height: 2em;
        border: 1px solid black;
        margin: 0 !important;
        border-radius: 0px !important;
        display: flex;
        }`
    });
  }

  get template() {
    return `
    <mat-paginator %[length]%
      %style%
      %class%
      %[pageSize]%
      %[pageSizeOptions]%
      %bCustomProps%
      %(page)%
    >
    </mat-paginator>
    `;
  }
  set template(templateString) {}
};
