// @jatahworx/bhive-toolkits/angular-material/component/Table/TableHeader.js
"use strict";
let AdvancedComponent = require("../../../core/AdvancedComponent");
let Attribute = require('../../../core/Attributes');

module.exports = class TableHeader extends AdvancedComponent {
  constructor() {
    const name = "ng-table-header";
    const designerTemplate = '<ng-table-header></ng-table-header>';
    const paletteTemplate = 'Header';
    const componentLabel = 'Header';

    super({
      name,
      designerTemplate,
      paletteTemplate,
      componentLabel,
      isAdvancedChild: true,
      templateUrl: "articles/#!components-guide-7/tables"
    });
    super.addAttribute(new Attribute({
      key: 'header',
      value: false,
      type: 'a',
      visible: false,
      isVisibleForParent: true
    }))
    super.composeTemplate();
  }

  fromParent(parentInstance) {
    this.shared = {};
    this.shared.displayColumns = parentInstance.shared.displayColumns;
  }

  get template() {
    let displayColumns = this.shared && this.shared.displayColumns ? this.shared.displayColumns : "";
    displayColumns = JSON.stringify(displayColumns).replace(new RegExp("\"", 'g'), "'");
    let headerAttribute = this.getHtmlAttributes(this.htmlAttributes);
    if (headerAttribute['header'] && headerAttribute['header']._value) {
      return `<tr mat-header-row *matHeaderRowDef="${displayColumns}"></tr>`;
    } else {
      return '';
    }
  }

  set template(t) {}
};
