// @jatahworx/bhive-toolkits/ServiceDesigner/ndefault-json/nodes/server/JSONConverter.js
let BaseComponent = require("@jatahworx/bhive-toolkits").BaseComponent;
let SRDUtil = require("@jatahworx/bhive-toolkits").SRDCodeGenUtil;

module.exports = class JSONConverter extends BaseComponent {
    constructor(constructorOptions) {
        super(constructorOptions.id, constructorOptions.name, "neutrinos_service_designer_JSONConvertor", "JSONConverter");

        let srdUtil = new SRDUtil();
        if (constructorOptions.inputMapping) {
            this.inputMapping = srdUtil.getValueFromTypedInput(constructorOptions.inputMapping);
        }
        if (constructorOptions.resultMapping) {
            this.resultMapping = srdUtil.getValueFromTypedInput(constructorOptions.resultMapping);
        }

        this.errorMapping = constructorOptions.errorMapping;
        this.viewType = 'server';
    }

    generateImports(rootPaths) {
        return [
            {
                library: rootPaths.pathToUtilsRoot + '/ndefault-json/JSONConverter/JSONConverterService',
                modules: ['JSONConverterService']
            }
        ]
    }

	generateSnippet() {
		let snippet = '';
		if (this.resultMapping) {
			snippet = `const p = JSONConverterService.getInstance();
				${this.resultMapping} = await p.json(${this.inputMapping});`;
		} else {
			snippet = `const p = JSONConverterService.getInstance();
				${this.inputMapping} = await p.json(${this.inputMapping});`;
		}
		return snippet;
	}



    get inputMapping() {
        return this.__inputMapping;
    }
    set inputMapping(inputMapping) {
        this.__inputMapping = inputMapping;
    }

    get resultMapping() {
        return this.__resultMapping;
    }

    set resultMapping(resultMapping) {
        this.__resultMapping = resultMapping;
    }

}
