// @jatahworx/bhive-toolkits/angular-material/component/AreaChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class AreaChart extends Component {
  constructor() {
    const name = 'NG-area-chart';
    const designerTemplate = `<div class=\"ad-charts ad-area-chart\" component-label="Area Chart"></div>`;
    const paletteTemplate = 'Area Chart';
    const componentLabel = 'Area Chart';
    const templateUrl =
      'articles/#!components-guide-7/ngx-area-chart';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    // not working sofar issue link https://github.com/swimlane/ngx-charts/issues/951
    // super.addAttribute(new Attribute("[activeEntries]", "", "kv", "p"));

    super.addAttribute(
      new Attribute({
        key: 'componentLabel',
        value: 'Area Chart',
        type: 'kv',
        useAsLabel: true
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[gradient]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legend]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showXAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showYAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(select)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[scheme]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[results]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[schemeType]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showGridLines]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legendTitle]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legendPosition]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: '[view]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[autoScale]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xScaleMin]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xScaleMax]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yScaleMin]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yScaleMax]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[animations]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[customColors]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[timeline]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[tooltipDisabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisTicks]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisTicks]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisTickFormatting]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisTickFormatting]',
        value: '',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NGX_CHARTS.val);
  }

  get template() {
    return `<div %fxFlex% %fxLayout%>
        <ngx-charts-area-chart %bCustomProps% %[scheme]% %[schemeType]% %[results]% %style% %class%  %[gradient]% %[view]% %[xAxis]% %[yAxis]% %[legend]% %[legendTitle]% %[legendPosition]% %[showXAxisLabel]% %[showYAxisLabel]% %[showGridLines]% %[yAxisLabel]% %[xAxisLabel]% %(select)% %[autoScale]% %[xScaleMin]% %[xScaleMax]% %[yScaleMin]% %[yScaleMax]% %[animations]% %[customColors]% %[timeline]% %[tooltipDisabled]% %[xAxisTicks]% %[yAxisTicks]% %[xAxisTickFormatting]% %[yAxisTickFormatting]%></ngx-charts-area-chart>
      </div>`;
  }
  set template(templateString) {}
};
