// @jatahworx/bhive-toolkits/angular-material/component/CardSubTitle.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class CardSubTitle extends Component {
  constructor() {
    const name = 'NG-CardSubTitle';
    const designerTemplate = `<div class=\"ad-card-sub-title\">
                                    <span class="component-placeholder">Card Sub-title</span>
                                  </div>`;
    const paletteTemplate = 'Card Sub-Title';
    const componentLabel = 'Card Sub-Title';
    const visible = false;
    const templateUrl = 'articles/#!components-guide/card5'; // 'articles/#!components-guide-7/card';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: 'Sub Title',
        value: 'Card Sub-Title',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'align',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-card-subtitle %bCustomProps% %style% %class% %align%>%Sub Title%</mat-card-subtitle>`;
    return template;
  }
  set template(templateString) {}
};
