// @jatahworx/bhive-toolkits/angular-material/component/CardActions.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class CardActions extends Component {
  constructor() {
    const name = 'NG-CardActions';
    const designerTemplate = `<div class=\"drop ad-flex\" component-label="Card Actions"></div>`;
    const paletteTemplate = 'Card Actions';
    const componentLabel = 'Card Actions';
    const visible = false;
    const templateUrl = 'articles/#!components-guide/card5'; // 'articles/#!components-guide-7/card';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-card-actions %bCustomProps% %style% %class%></mat-card-actions>`;
    return template;
  }
  set template(templateString) {}
};
