// @jatahworx/bhive-toolkits/angular-material/component/CardImage.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class CardImage extends Component {
  constructor() {
    const name = 'NG-CardImage';
    const designerTemplate = `<div class=\"ad-card-image\" component-label="Card Image"></div>`;
    const paletteTemplate = 'Card Image';
    const componentLabel = 'Card Image';
    const visible = false;
    const templateUrl = 'articles/#!components-guide/card5'; // 'articles/#!components-guide-7/card';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'alt',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'imgSrc',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'height',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<img mat-card-image %bCustomProps% %style% %class% %alt% %imgSrc% %height%>`;
    return template;
  }
  set template(templateString) {}
};
