// @jatahworx/bhive-toolkits/angular-material/component/CardHeader.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class CardHeader extends Component {
  constructor() {
    const name = 'NG-CardHeader';
    const designerTemplate = ` <div class=\"ad-card-header\" component-label="Card Header">
                                    <div class=\"ad-card-avatar\"></div>
                                    <div class=\"ad-card-header-title\">Title</div>
                                    <div class=\"ad-card-subtitle\">Subtitle</div>
                                    </div>`;
    const paletteTemplate = 'Card Header';
    const componentLabel = 'Card Header';
    const visible = false;
    const templateUrl = 'articles/#!components-guide/card5'; // 'articles/#!components-guide-7/card';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'title',
      value: '',
      type: 'a'
    }));
    super.addAttribute(
      new Attribute({
        key: 'subtitle',
        value: '',
        type: 'a'
      })
    );
    super.addAttribute(new Attribute({
      key: 'imgSrc',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'alt',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `
    <mat-card-header %bCustomProps% %style% %class%>
      <mat-card-title>%title%</mat-card-title>
      <mat-card-subtitle>%subtitle%</mat-card-subtitle>
      <img mat-card-avatar %imgSrc% %alt%>
    </mat-card-header>`;
    return template;
  }
  set template(templateString) {}
};
