// @jatahworx/bhive-toolkits/angular-material/component/ChipList.js
"use strict";
let AdvancedComponent = require("../../core/AdvancedComponent");
let Attribute = require("../../core/Attributes");

module.exports = class ChipList extends AdvancedComponent {
    constructor() {
        const name = 'n-chip-list';
        const designerTemplate = `<n-chip-list onclick="click(event)" class="ad-element flex-shrink-0 flex-grow-1" component-label="Chips">
                                    <n-chip n-adv-child></n-chip>
                                    <div slot="add" class="ad-card-buttons">
                                      <div class="inline-block">
                                        <button id="addchip" class="add-child-button flex-column">Add Chip</button>
                                      </div>
                                    </div>
                                  </n-chip-list>`
        const paletteTemplate = 'Chips';
        const componentLabel = 'Chips';
        const visible = true;
        const templateUrl = 'articles/#!components-guide-7/chips';

        super({
            name,
            designerTemplate,
            template: ``,
            componentLabel,
            paletteTemplate,
            visible,
            templateUrl
        });

        super.setType(AdvancedComponent.COMPONENT_TYPE_TITLES.LAYOUT.val);

        super.addAttribute(
            new Attribute({
              key: '[selectable]',
              value: 'true',
              type: 'kv'
            })
        );

        super.addAttribute(
            new Attribute({
              key: '[multiple]',
              value: 'true',
              type: 'kv'
            })
        );

        super.addAttribute(
            new Attribute({
              key: '(change)',
              value: '',
              type: 'kv',
              complexity: 'advanced'
            })
        );

        super.addAttribute(
          new Attribute({
            key: '[compareWith]',
            value: '',
            type: 'kv',
            complexity: 'advanced'
          })
        );

        super.composeTemplate({
            styles: `
            :host {
              display: flex;
              padding-top: 2em;
              padding-bottom: 1em;
              min-width: 20em;
              flex-direction: column;
              align-self: start;
            }
           .parent_card {
             display: flex;
             flex-direction: row;
          }
          .chip-margin {
            margin-left: 0.8em;
          }
            `,
            slotsTemplate: `   
            <div class="parent_card chip-margin">
              <slot name="chips"></slot>
            </div>
            <div class="parent_card">
              <slot id="addSlot" name="add"></slot>
            </div>
            `,
            onInit: function () {},
            onDestroy: function () {},
            customMethods: {
              click: function (e) {
                if (e.target.id === "addchip") {
                  this.addChild("n-chip");
                }
              }
            }
          });

    }

    get template() {
        const template = `
        <mat-chip-list
          %style%
          %class%
          %[selectable]%
          %[multiple]% 
          %(change)%
          %[compareWith]% 
          %bCustomProps%>
        </mat-chip-list>`;
        return template;
    }
    set template(templateString) {}
}
