// @jatahworx/bhive-toolkits/angular-material/component/CustomHtml.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class CustomHtml extends Component {
  constructor() {
    const name = 'HTML';
    const designerTemplate = `<div class="custom-html" b-props="html-editor" component-label="HTML">
                                    <i class=\"material-icons\">code</i>
                                  </div>`;
    const paletteTemplate = 'HTML';
    const componentLabel = 'HTML';
    const templateUrl = '';
    const visible = false;
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'html',
      value: '',
      type: 'a'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.OTHERS.val);
  }
  get template() {
    const template = '%html%';
    return template;
  }
  set template(templateString) {}
};
