// @jatahworx/bhive-toolkits/angular-material/component/ExpansionHeader.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ExpansionPanelHeader extends Component {
  constructor() {
    const name = 'NG-ExpansionPanelHeader';
    const designerTemplate = `<div class=\"drop\" component-label="Expansion Header"></div>`;
    const paletteTemplate = 'Expansion Header';
    const componentLabel = 'Expansion Header';
    const templateUrl =
      'articles/#!components-guide-7/expansion-panel-header';
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: '[collapsedHeight]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[expandedHeight]',
        value: '',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-expansion-panel-header %style% %bCustomProps% %class% %[collapsedHeight]% %[expandedHeight]%>
    </mat-expansion-panel-header>`;
    return template;
  }
  set template(templateString) {}
};
