// @jatahworx/bhive-toolkits/angular-material/component/ExpansionDesc.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ExpansionPanelDesc extends Component {
  constructor() {
    const name = 'NG-ExpansionPanelDesc';
    const designerTemplate = `<div class=\"drop\" component-label="Expansion Description"></div>`;
    const paletteTemplate = 'Expansion Description';
    const componentLabel = 'Expansion Description';
    const templateUrl =
      'articles/#!components-guide-7/expansion-panel-description';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: 'componentLabel',
        value: 'Expansion Description',
        type: 'kv',
        useAsLabel: true
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'description',
        value: '',
        type: 'a'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `
    <mat-panel-description %style% %bCustomProps% %class%>
      %description%
    </mat-panel-description>`;
  return template;
  }
  set template(templateString) {}
};
