// @jatahworx/bhive-toolkits/angular-material/component/ExpansionAccordion.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ExpansionAccordion extends Component {
  constructor() {
    const name = 'NG-ExpansionAccordion';
    const designerTemplate = `<div class=\"drop \" component-label="Expansion Panel Outlet"></div>`;
    const paletteTemplate = 'Expansion Panel Outlet';
    const componentLabel = 'Expansion Panel Outlet';

    const templateUrl =
      'articles/#!components-guide-7/expansion-panel-outlet';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );

    super.addAttribute(
      new Attribute({
        key: '[displayMode]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'multi',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-accordion %style% %bCustomProps% %class% %[displayMode]% %multi% >
    </mat-accordion>`;
    return template;
  }
  set template(templateString) {}
};
