// @jatahworx/bhive-toolkits/angular-material/component/ExpansionTitle.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ExpansionPanelTitle extends Component {
  constructor() {
    const name = 'NG-ExpansionPanelTitle';
    const designerTemplate = `<div class=\"drop\" component-label="Expansion Title"></div>`;
    const paletteTemplate = 'Expansion Title';
    const componentLabel = 'Expansion Title';
    const templateUrl =
      'articles/#!components-guide-7/expansion-panel-title';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'title',
      value: '',
      type: 'a'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `
    <mat-panel-title %style% %bCustomProps% %class%>
      %title%
    </mat-panel-title>`;
    return template;
  }
  set template(templateString) {}
};
