// @jatahworx/bhive-toolkits/angular-material/component/FabButton.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class FabButton extends Component {
  constructor() {
    const name = 'NG-FabButton';
    const designerTemplate = `<div class=\"ad-fab non-dropable-comp-border form-control-comp\">
    <div class=\"fab-icon fab-icons-container\"> <i class="material-icons component-placeholder">info</i></div>
                                  </div>`;
    const paletteTemplate = 'fab Button';
    const componentLabel = 'fab Button';
    const templateUrl =
      'articles/#!components-guide-7/fab-button';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
     
    );

    super.addAttribute(
      new Attribute({
        key: 'fabicon',
        value: 'info',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '(click)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disableRipple]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'type',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'routerLink',
        value: '',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
  }
  get template() {
    const template = `<button mat-fab %style% %class% %color% %bCustomProps% %routerLink% %(click)% %[disabled]% %[disableRipple]% %type%><mat-icon>%fabicon%</mat-icon></button>`;
    return template;
  }
  set template(templateString) {}
};
