// @jatahworx/bhive-toolkits/angular-material/component/FlexLayoutRow.js
'use strict';
let Component = require('../../core/Component');
let Attribute = require('../../core/Attributes');

module.exports = class FlexLayoutRow extends Component {
  constructor() {
    const name = 'FlexLayoutRow';
    const designerTemplate = `<div class="ad-flex drop" component-label="Row"></div>`;
    const paletteTemplate = 'Row';
    const componentLabel = 'Row';
    const templateUrl =
      'articles/#!components-guide-7/flex-layout-row';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'fxLayoutGap',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'wrap',
        value: 'wrap',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'layoutAlign',
        value: 'start',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'pLayoutAlign',
        value: 'stretch',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[fxShow]',
        value: 'true',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[fxHide]',
        value: 'false',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<div fxLayout=\"row %wrap%\" %style% %class% %fxFlex%  %fxLayoutGap% fxLayoutAlign = \"%layoutAlign% %pLayoutAlign%\" %[fxShow]% %[fxHide]% %bCustomProps%></div>`;
    return template;
  }
  set template(templateString) {}
};
