// @jatahworx/bhive-toolkits/angular-material/component/FlexLayoutCol.js
'use strict';
let Component = require('../../core/Component');
let Attribute = require('../../core/Attributes');

module.exports = class FlexLayoutCol extends Component {
  constructor() {
    const name = 'FlexLayoutCol';
    const designerTemplate = `<div class=\"drop\" component-label="Column"></div>`;
    const paletteTemplate = 'Column';
    const componentLabel = 'Column';
    const templateUrl =
      'articles/#!components-guide-7/flex-layout-column';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'fxLayoutGap',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'wrap',
        value: 'wrap',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'layoutAlign',
        value: 'start',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'pLayoutAlign',
        value: 'stretch',
        type: 'a'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[fxShow]',
        value: 'true',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[fxHide]',
        value: 'false',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<div fxLayout=\"column %wrap%\" %style% %class% %fxFlex%  %fxLayoutGap% fxLayoutAlign = \"%layoutAlign% %pLayoutAlign%\" %[fxShow]% %[fxHide]% %bCustomProps%></div>`;
    return template;
  }
  set template(templateString) {}
};
