// @jatahworx/bhive-toolkits/angular-material/component/GoogleMap.js

'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

/** 
*  THIS COMPONENT IS DEPRECATED FROM VERSION: 7.8.0
**/
module.exports = class GoogleMap extends Component {
  constructor() {
    const name = 'NG-GoogleMap';
    const designerTemplate = `<div class=\"ad-google-map non-dropable-comp-border\" component-label="File Upload">
                                <div class=\"google-map-icon height-100\"></div>
                                  </div>`;
    const paletteTemplate = 'Google Map';
    const componentLabel = 'Google Map';
    const templateUrl = 'https://angular-maps.com/';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: false
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: '[iconUrl]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[hideMarker]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[enableHighAccuracy]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[mapTimedout]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[mapZoom]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[markerSize]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[usePanning]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: '[lat]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '[lng]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[markerTraggerble]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(gMapClick)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(gMarkerClick)',
        value: '',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.OTHERS.val);
  }
  get template() {
    const template = `<n-google-map style=\"width:100%;height:100%\" %bCustomProps% %[hideMarker]% %[iconUrl]% %[lat]% %[lng]% %[enableHighAccuracy]% %[mapTimedout]% %[mapZoom]% %[markerSize]% %[usePanning]% %[markerTraggerble]% %(gMapClick)% %(gMarkerClick)%></n-google-map>`;
    return template;
  }
  set template(templateString) {}
};
