// @jatahworx/bhive-toolkits/angular-material/component/GridTile.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class GridTile extends Component {
  constructor() {
    const name = 'NG-GridTile';
    const designerTemplate = `<div class=\"drop\" component-label="Grid Tile"></div>`;
    const paletteTemplate = 'Grid Tile';
    const componentLabel = 'Grid Tile';
    const templateUrl =
      'articles/#!components-guide-7/grid-tile';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(new Attribute({
      key: '*ngFor',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[rowspan]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[colspan]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'label',
      value: '',
      type: 'a'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-grid-tile %*ngFor% %[rowspan]% %[colspan]% %bCustomProps% %style% %class%>%label%</mat-grid-tile>`;
    return template;
  }
  set template(templateString) {}
};
