// @jatahworx/bhive-toolkits/angular-material/component/GridList.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class GridList extends Component {
  constructor() {
    const name = 'NG-GridList';
    const designerTemplate = `<div class=\"drop\" component-label="Grid List"></div>`;
    const paletteTemplate = 'Grid List';
    const componentLabel = 'Grid List';
    const templateUrl =
      'articles/#!components-guide-7/grid-list';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
      
    );

    super.addAttribute(new Attribute({
      key: 'cols',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'gutterSize',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'rowHeight',
        value: '',
        type: 'kv'
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);
  }
  get template() {
    const template = `<mat-grid-list %rowHeight% %cols% %gutterSize% %bCustomProps% %style% %class%></mat-grid-list>`;
    return template;
  }
  set template(templateString) {}
};
