// @jatahworx/bhive-toolkits/angular-material/component/H1.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class H1 extends Component {
  constructor() {
    const name = 'NG-H1';
    const designerTemplate = `<div class=\"ad-header-elements ad-h1\">
                                    <span class="component-placeholder">h1</span>
                                  </div>`;
    const paletteTemplate = 'h1';
    const componentLabel = 'h1';
    const templateUrl = 'articles/#!components-guide/h1'; // 'articles/#!components-guide-7/html';
    const visible = false;
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    });

    super.addAttribute(
      new Attribute({
        key: 'H1 Text',
        value: '',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'align',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
  }
  get template() {
    const template = `<h1 %bCustomProps% %style% %class% %align%>%H1 Text%</h1>`;
    return template;
  }
  set template(templateString) {}
};
