// @jatahworx/bhive-toolkits/angular-material/component/H2.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class H2 extends Component {
  constructor() {
    const name = 'NG-H2';
    const designerTemplate = `<div class=\"ad-header-elements ad-h2\">
                                    <span class="component-placeholder">h2</span>
                                  </div>`;
    const paletteTemplate = 'h2';
    const componentLabel = 'h2';
    const templateUrl =  'articles/#!components-guide/h2'; // 'articles/#!components-guide-7/html';
    const visible = false;

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: 'H2 Text',
        value: '',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'align',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
  }
  get template() {
    const template = `<h2 %bCustomProps% %style% %class% %align%>%H2 Text%</h2>`;
    return template;
  }
  set template(templateString) {}
};
