// @jatahworx/bhive-toolkits/angular-material/component/H3.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class H3 extends Component {
  constructor() {
    const name = 'NG-H3';
    const designerTemplate = `<div class=\"ad-header-elements ad-h3\">
                                    <span class="component-placeholder">h3</span>                                    
                                </div>`;
    const paletteTemplate = 'h3';
    const componentLabel = 'h3';
    const templateUrl = 'articles/#!components-guide/h3';  //'articles/#!components-guide-7/html';
    const visible = false;

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
      
    );

    super.addAttribute(
      new Attribute({
        key: 'H3 Text',
        value: '',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'align',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
  }
  get template() {
    const template = `<h3 %bCustomProps% %style% %class% %align%>%H3 Text%</h3>`;
    return template;
  }
  set template(templateString) {}
};
