// @jatahworx/bhive-toolkits/angular-material/component/HorizontalBarChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class HorizontalBarChart extends Component {
  constructor() {
    const name = 'NG-bar-chart-horizontal';
    const designerTemplate = `<div class=\"ad-horizontal-bar-chart ad-charts-align ad-charts\" component-label="Horizontal Bar Chart"></div>`;
    const paletteTemplate = 'Horizontal Bar Chart';
    const componentLabel = 'Horizontal Bar Chart';
    const templateUrl =
      'articles/#!components-guide-7/ngx-stacked-horizontal-bar-chart';
 
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
     
    );

    super.addAttribute(
      new Attribute({
        key: '[gradient]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legend]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showXAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showYAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(select)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[scheme]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[results]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
    super.setType(Component.COMPONENT_TYPE_TITLES.NGX_CHARTS.val);
  }
  get template() {
    const template = `
    <div %fxLayout% %fxFlex%>
     <ngx-charts-bar-horizontal %bCustomProps% %[scheme]% %[results]% %style% %class% %[gradient]% %[xAxis]% %[yAxis]% %[legend]% %[showXAxisLabel]% %[showYAxisLabel]% %[yAxisLabel]% %[xAxisLabel]% %(select)%></ngx-charts-bar-horizontal>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
