// @jatahworx/bhive-toolkits/angular-material/component/Image.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class Image extends Component {
  constructor() {
    const name = 'NG-image';
    const designerTemplate = `<div class=\"ad-image-component ad-image ad-image-align\" component-label="Image"></div>`;
    const paletteTemplate = 'Image';
    const componentLabel = 'Image';
    const templateUrl = 'articles/#!components-guide-7/image';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });

    super.addAttribute(new Attribute({
      key: 'imgSrc',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'alt',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '[src]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'imgSecure',
        value: '',
        type: 'a'
      })
    );
    super.addAttribute(new Attribute({
      key: 'width',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'srcset',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'sizes',
      value: '',
      type: 'kv'
    }));

    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
  }
  get template() {
    const template = `<img %imgSrc% %[src]% %alt% %style% %bCustomProps% %class% %imgSecure% %width% %srcset% %sizes%>`;
    return template;
  }
  set template(templateString) {}
};
