// @jatahworx/bhive-toolkits/angular-material/component/LineChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class LineChart extends Component {
  constructor() {
    const name = 'NG-line-chart';
    const designerTemplate = `<div class=\"ad-charts ad-line-chart\" component-label="Line Chart"></div>`;
    const paletteTemplate = 'Line Chart';
    const componentLabel = 'Line Chart';
    const templateUrl =
      'articles/#!components-guide-7/ngx-line-chart';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NGX_CHARTS.val);
    // not working sofar issue link https://github.com/swimlane/ngx-charts/issues/951
    // super.addAttribute(new Attribute("[activeEntries]", "", "kv", "p"));
    // not working issue link https://github.com/swimlane/ngx-charts/issues/1080
    // super.addAttribute(new Attribute("[legendPosition]", "", "kv", "p"));
    super.addAttribute(
      new Attribute({
        key: '[gradient]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legend]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showXAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showYAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(select)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[scheme]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[results]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: '[view]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[schemeType]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[customColors]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[animations]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[rangeFillOpacity]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legendTitle]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showGridLines]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[roundDomains]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisTicks]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisTicks]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisTickFormatting]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisTickFormatting]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[timeline]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[autoScale]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[tooltipDisabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[referenceLines]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showRefLines]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showRefLabels]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xScaleMin]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xScaleMax]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yScaleMin]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yScaleMax]',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `
    <div %fxFlex% %fxLayout%>
     <ngx-charts-line-chart %bCustomProps% %[scheme]% %[results]%  %style% %class%  %[gradient]% %[xAxis]% %[yAxis]% %[legend]% %[showXAxisLabel]% %[showYAxisLabel]% %[yAxisLabel]% %[xAxisLabel]% %(select)% %[view]% %[schemeType]% %[customColors]% %[animations]% %[rangeFillOpacity]% %[legendTitle]% %[showGridLines]% %[roundDomains]% %[xAxisTicks]% %[yAxisTicks]% %[xAxisTickFormatting]% %[yAxisTickFormatting]% %[timeline]% %[autoScale]% %[tooltipDisabled]% %[referenceLines]% %[showRefLines]% %[showRefLabels]% %[xScaleMin]% %[xScaleMax]% %[yScaleMin]% %[yScaleMax]%></ngx-charts-line-chart>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
