// @jatahworx/bhive-toolkits/angular-material/component/ListItem.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ListItem extends Component {
  constructor() {
    const name = 'NG-listItem';
    const designerTemplate = `<div class=\"drop\" component-label="List Item"></div>`;
    const paletteTemplate = 'List Item';
    const componentLabel = 'List Item';
    const templateUrl = 'articles/#!components-guide-7/list-item';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);

    super.addAttribute(new Attribute({
      key: '*ngFor',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'label',
      value: '',
      type: 'a'
    }));
  }
  get template() {
    const template = `<mat-list-item %*ngFor%  %style% %bCustomProps% %class%>%label%</mat-list-item>`;
    return template;
  }
  set template(templateString) {}
};
