// @jatahworx/bhive-toolkits/angular-material/component/MenuButton.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class MenuButton extends Component {
  constructor() {
    const name = 'NG-MenuButton';
    const designerTemplate = `<div class=\"ad-button non-dropable-comp-border form-control-comp\">
                                    <span class="component-placeholder">Menu Button</span>
                                  </div>`;
    const paletteTemplate = 'Menu Button';
    const componentLabel = 'Menu Button';
    const templateUrl = 'articles/#!components-guide-7/menu-button';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
  });
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(
      new Attribute({
        key: 'MenuName',
        value: '',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[matMenuTriggerFor]',
        value: 'menu',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<button mat-button %style% %class% %bCustomProps% %color% %[matMenuTriggerFor]%>%MenuName%</button>`;
    return template;
  }
  set template(templateString) {}
};
