// @jatahworx/bhive-toolkits/angular-material/component/MenuItem.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class MenuItem extends Component {
  constructor() {
    const name = 'NG-MenuItem';
    const designerTemplate = `<div class=\"drop ad-flex\" component-label="Menu Items"></div>`;
    const paletteTemplate = 'Menu Item';
    const componentLabel = 'Menu Item';
    const templateUrl = 'articles/#!components-guide-7/menu-item';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(
      new Attribute({
        key: 'MenuItemName',
        value: '',
        type: 'a'
      })
    );

    super.addAttribute(
      new Attribute({
        key: '(click)',
        value: '',
        type: 'kv'
      })
    );
    
    super.addAttribute(
      new Attribute({
        key: '[disabled]',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<button mat-menu-item %bCustomProps% %style% %class% %[disabled]% %(click)%> %MenuItemName% </button>`;
    return template;
  }
  set template(templateString) {}
};
