// @jatahworx/bhive-toolkits/angular-material/component/NgBubbleChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class NgBubbleChart extends Component {
  constructor() {
    const name = 'NG-ng-bubble-chart';
    const designerTemplate = `<div class=\"ad-charts  ad-bubble-chart\" component-label="Bubble Chart"></div>`;
    const paletteTemplate = 'Bubble Chart';
    const componentLabel = 'Bubble Chart';
    const templateUrl = 'articles/#!components-guide-7/ng-bubble-chart';
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });

    super.setType(Component.COMPONENT_TYPE_TITLES.NG_CHARTS.val);

    super.addAttribute(
      new Attribute({
        key: '[datasets]',
        value: '100',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[options]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '100',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[labels]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[colors]',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `
    <div %fxFlex% %fxLayout%>
      <canvas baseChart %[datasets]% %bCustomProps% %style% %class% %[options]% %[labels]% %[colors]% chartType=\"bubble\"></canvas>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
