// @jatahworx/bhive-toolkits/angular-material/component/NgDoughnutChart.js
"use strict";
let Attribute = require("../../core/Attributes");
let Component = require("../../core/Component");
/**
 * Previously horizontal bar chart will now be displayed
 * as bar chart before of ng2-charts changes
 */
module.exports = class NgDoughnutChart extends Component {
  constructor() {
    const name = "NG-ng-doughnut-chart";
    const designerTemplate = `<div class=\"ad-doughnut-chart  ad-charts\" component-label="Doughnut Chart"></div>`;
    const paletteTemplate = "Doughnut Chart";
    const componentLabel = "Doughnut Chart";
    const templateUrl =
      "articles/#!components-guide-7/ng-doughnut-chart";

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NG_CHARTS.val);

    super.addAttribute(
      new Attribute({
        key: "[data]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[datasets]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[labels]",
        value: "100",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[options]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[colors]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[legend]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartHover)",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartClick)",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxLayout",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxFlex",
        value: "",
        type: "kv"
      })
    );
  }
  get template() {
    const template = `
    <div %fxFlex% %fxLayout%>
      <canvas baseChart %[data]% %[datasets]% %bCustomProps% %style% %class% %[labels]% %[options]% %[colors]% %[legend]% chartType=\"doughnut\" %(chartHover)% %(chartClick)%></canvas>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
