// @jatahworx/bhive-toolkits/angular-material/component/NgLineChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class NgLineChart extends Component {
  constructor() {
    const name = 'NG-ng-line-chart';
    const designerTemplate = `<div class=\"ad-charts  ad-line-chart\" component-label="Line Chart"></div>`;
    const paletteTemplate = 'Line Chart';
    const componentLabel = 'Line Chart';
    const templateUrl = 'articles/#!components-guide-7/ng-line-chart';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NG_CHARTS.val);

    super.addAttribute(new Attribute({
      key: '[data]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[datasets]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[labels]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[options]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[colors]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legend]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(chartHover)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(chartClick)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '100',
      type: 'kv'
    }));
  }
  get template() {
    const template = `
    <div %fxFlex% %fxLayout%>
      <canvas baseChart %[data]% %[datasets]% %bCustomProps% %style% %class% %[labels]% %[options]% %[colors]% %[legend]% chartType=\"line\" %(chartHover)% %(chartClick)%></canvas>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
