// @jatahworx/bhive-toolkits/angular-material/component/NgPieChart.js
"use strict";
let Attribute = require("../../core/Attributes");
let Component = require("../../core/Component");

module.exports = class NgPieChart extends Component {
  constructor() {
    const name = "NG-ng-pie-chart";
    const designerTemplate = `<div class=\"ad-pie-chart  ad-charts\" component-label="Pie Chart"></div>`;
    const paletteTemplate = "Pie Chart";
    const componentLabel = "Pie Chart";
    const templateUrl =
      "articles/#!components-guide-7/ng-pie-chart";

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NG_CHARTS.val);

    super.addAttribute(
      new Attribute({
        key: "[data]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[datasets]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[labels]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[options]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[colors]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[legend]",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartHover)",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartClick)",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxLayout",
        value: "",
        type: "kv"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxFlex",
        value: "100",
        type: "kv"
      })
    );
  }
  get template() {
    const template = `
    <div %fxFlex% %fxLayout%>
      <canvas baseChart %bCustomProps% %style% %class% %[data]% %[datasets]% %[labels]% %[options]% %[colors]% %[legend]% chartType=\"pie\" %(chartHover)% %(chartClick)%></canvas>
    </div>`;
    return template;
  }
  set template(templateString) {}
};
