// @jatahworx/bhive-toolkits/angular-material/component/NgRadarChart.js
"use strict";
let Attribute = require("../../core/Attributes");
let Component = require("../../core/Component");

module.exports = class NgRadarChart extends Component {
  constructor() {
    const name = "NG-ng-radar-chart";
    const designerTemplate = `<div class=\"ad-radar-chart  ad-charts\" component-label="Radar Chart"></div>`;
    const paletteTemplate = "Radar Chart";
    const componentLabel = "Radar Chart";
    const templateUrl =
      "articles/#!components-guide-7/ng-radar-chart";

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.addAttribute(
      new Attribute({
        key: "[data]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[datasets]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[labels]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[options]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[colors]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "[legend]",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartHover)",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "(chartClick)",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxLayout",
        value: "",
        type: "kv",
        attrDef: "p"
      })
    );
    super.addAttribute(
      new Attribute({
        key: "fxFlex",
        value: "100",
        type: "kv",
        attrDef: "p"
      })
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NG_CHARTS.val);
  }
  get template() {
    const template = `<div %fxFlex% %fxLayout%>
        <canvas baseChart %[data]% %[datasets]%  %bCustomProps% %style% %class% %[labels]% %[options]% %[colors]% %[legend]% chartType=\"radar\" %(chartHover)% %(chartClick)%></canvas>
      </div>`;
    return template;
  }
  set template(templateString) {}
};
