// @jatahworx/bhive-toolkits/angular-material/component/Paragraph.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class Paragraph extends Component {
  constructor() {
    const name = 'NG-Paragraph';
    const designerTemplate = `<div class=\"ad-paragraph ad-h3\">
                                    <span class="component-placeholder align-text-left">Paragraph</span>
                                  </div>`;
    const paletteTemplate = 'Paragraph';
    const componentLabel = 'Paragraph';
    const templateUrl = 'articles/#!components-guide/paragraph5'; //'articles/#!components-guide-7/paragraph';
    const visible = false;

    super({
      name : name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);

    super.addAttribute(
      new Attribute({
        key: 'text',
        value: '',
        type: 'a',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: 'align',
      value: '',
      type: 'kv'
    }));
  }
  get template() {
    const template = `<p %bCustomProps% %style% %class% %align%>%text%</p>`;
    return template;
  }
  set template(templateString) {}
};
