// @jatahworx/bhive-toolkits/angular-material/component/ProgressSpinner.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ProgressSpinner extends Component {
  constructor() {
    const name = 'NG-ProgressSpinner';
    const designerTemplate = `<div class=\"non-dropable-comp-border form-control-comp\">
                                <div class="progress-spinner-icon icon-size height-100"></div>
                              </div>`;
    const paletteTemplate = 'Progress Spinner';
    const componentLabel = 'Progress Spinner';
    const templateUrl =
      'articles/#!components-guide-7/progress-spinner';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
    super.addAttribute(
      new Attribute({
        key: 'componentLabel',
        value: 'Progress Spinner',
        type: 'kv',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: '[progressSpinnerMode]',
      value: '',
      type: 'a'
    }));
    super.addAttribute(
      new Attribute({
        key: 'strokeWidth',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[value]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[diameter]',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<mat-progress-spinner %style% %class% %bCustomProps% [mode]=\"%[progressSpinnerMode]%\" %color% %[value]% %strokeWidth% %[diameter]%></mat-progress-spinner>`;
    return template;
  }
  set template(templateString) { }
};
