// @jatahworx/bhive-toolkits/angular-material/component/ProgressBar.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class ProgressBar extends Component {
  constructor() {
    const name = 'NG-ProgressBar';
    const designerTemplate = `<div class=\"progress-bar-wrapper\" component-label="Progress Bar">
    <div class="label-padding"></div>
    <div class="progress-bar-icon-container">
      <div class="progress-bar-icon"></div>
    </div>
  </div>`;
    const paletteTemplate = 'Progress Bar';
    const componentLabel = 'Progress Bar';
    const templateUrl =
      'articles/#!components-guide-7/progress-bar';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);
    super.addAttribute(new Attribute({
      key: '[progressBarMode]',
      value: '',
      type: 'a'
    }));
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'value',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'bufferValue',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(animationEnd)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[progressbarId]',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<mat-progress-bar %style% %class% %bCustomProps% [mode]=\"%[progressBarMode]%\" %color% %value% %bufferValue% %(animationEnd)% %[progressbarId]%></mat-progress-bar>`;
    return template;
  }
  set template(templateString) { }
};
