// @jatahworx/bhive-toolkits/angular-material/component/Radio.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class Radio extends Component {
  constructor() {
    const name = 'NG-Radio';
    const designerTemplate = `<div class=\"ad-radio non-dropable-comp-border form-control-comp\">
                                    <i class="material-icons ad-radio-icon-margin">radio_button_checked</i>
                                  </div>`;
    const paletteTemplate = 'Radio Button';
    const componentLabel = 'Radio Button';
    const templateUrl = 'articles/#!components-guide-7/radio-button';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);

    super.addAttribute(
      new Attribute({
        key: 'componentLabel',
        value: 'Radio Button',
        type: 'kv',
        useAsLabel: true
      })
    );
    super.addAttribute(new Attribute({
      key: '[value]',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[required]',
        value: 'false',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'label',
      value: '',
      type: 'a'
    }));
    super.addAttribute(new Attribute({
      key: 'id',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'name',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[checked]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(change)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[labelPosition]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disableRipple]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
  }
  get template() {
    const template = `<mat-radio-button %[value]% %style% %class% %[required]% %bCustomProps% %id% %name% %(change)% %[disabled]% %[disableRipple]% %[labelPosition]% %[checked]% %color%>%label%</mat-radio-button>`;
    return template;
  }
  set template(templateString) { }
};
