// @jatahworx/bhive-toolkits/angular-material/component/RaisedButton.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class RaisedButton extends Component {
  constructor() {
    const name = 'NG-RaisedButton';
    const designerTemplate = `<div class=\"ad-raised-button non-dropable-comp-border form-control-comp\">
                                    <span class="component-placeholder">Raised Button</span>
                                  </div>`;
    const paletteTemplate = 'Raised Button';
    const componentLabel = 'Raised Button';
    const templateUrl =
      'articles/#!components-guide-7/rised-button';

    super({
      name: name,
      designerTemplate:designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel:componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.FORM_CONTROLS.val);

    super.addAttribute(
      new Attribute({
        key: 'buttonname',
        value: '',
        type: 'a',
        useAsLabel: true,
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '(click)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[disableRipple]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'type',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'routerLink',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<button mat-raised-button %style%  %routerLink% %class% %bCustomProps% %color% %(click)% %[disabled]% %[disableRipple]% %type%>%buttonname%</button>`;
    return template;
  }
  set template(templateString) {}
};
