// @jatahworx/bhive-toolkits/angular-material/component/RouterOutlet.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class RouterOutlet extends Component {
  constructor() {
    const name = 'NG-RouterOutlet';
    const designerTemplate = `<div class=\"router-icon routerborder ad-image-align\" component-label="Router Outlet"></div>`;
    const paletteTemplate = 'Router Outlet';
    const componentLabel = 'Router Outlet';
    const templateUrl =
      'articles/#!components-guide-7/router-outlet';
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      addDefaultAttr: false
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(
      new Attribute({
        key: '(activate)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(deactivate)',
        value: '',
        type: 'kv'
      })
    );
    
  }
  get template() {
    const template = `<router-outlet %bCustomProps% %(activate)% %(deactivate)%></router-outlet>`;
    return template;
  }
  set template(templateString) {}
};
