// @jatahworx/bhive-toolkits/angular-material/component/SideNavContainer.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class SideNavContainer extends Component {
  constructor() {
    const name = 'NG-sidenav-container';
    const designerTemplate = `<div class=\"ad-sidenav-container drop\" component-label="Sidenav Container">
                                  </div>`;
    const paletteTemplate = 'Sidenav Container';
    const componentLabel = 'Sidenav Container';
    const templateUrl =
      'articles/#!components-guide-7/sidenav-container';
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    });
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(
      new Attribute({
        key: 'autosize',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[hasBackdrop]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(backdropClick)',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<mat-sidenav-container %style% %bCustomProps% %class% %autosize% %[hasBackdrop]% %(backdropClick)%></mat-sidenav-container>`;
    return template;
  }
  set template(templateString) {}
};
