// @jatahworx/bhive-toolkits/angular-material/component/Step.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class Step extends Component {
  constructor() {
    const name = 'NG-Step';
    const designerTemplate = `<div class=\"drop ad-flex\" component-label="Step"></div>`;
    const paletteTemplate = 'Step';
    const componentLabel = 'Step';
    const templateUrl =
      'articles/#!components-guide-7/step';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.LAYOUT.val);

    super.addAttribute(new Attribute({
      key: 'label',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'completed',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'editable',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'optional',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[stepControl]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'state',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'hasError',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'errorMessage',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<mat-step %label% %completed% %[stepControl]% %optional% %editable% %bCustomProps% %style% %class% %state% %hasError% %errorMessage%></mat-step>`;
    return template;
  }
  set template(templateString) {}
};
