// @jatahworx/bhive-toolkits/angular-material/component/Tab.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');
module.exports = class Tab extends Component {
  constructor() {
    const name = 'NG-tab';
    const designerTemplate = `<div class=\"drop\" component-label="Tab"></div>`;
    const paletteTemplate = 'Tab';
    const visible = false;
    const componentLabel = 'Tab';
    const templateUrl =  'articles/#!components-guide/tab5'; //'articles/#!components-guide-7/tab';
    super({
      name:name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible: visible
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(new Attribute({
      key: 'label',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: '[disabled]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[isActive]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'position',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'index',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'origin',
      value: '',
      type: 'kv'
    }));
  }
  get template() {
    const template = `<mat-tab %style% %class% %label% %bCustomProps% %[disabled]% %[isActive]% %position% %index% %origin%></mat-tab>`;
    return template;
  }
  set template(templateString) {}
};
