// @jatahworx/bhive-toolkits/angular-material/component/TabGroup.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');
module.exports = class TabGroup extends Component {
  constructor() {
    const name = 'NG-TabGroup';
    const designerTemplate = `<div class="ad-flex drop" component-label="Tab Group"></div>`;
    const paletteTemplate = 'Tab Group';
    const visible = false;
    const componentLabel = 'Tab Group';
    const templateUrl = 'articles/#!components-guide/tab-group5'; // 'articles/#!components-guide-7/tab-group';
    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl,
      visible:visible
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(
      new Attribute({
        key: 'animationDuration',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'backgroundColor',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(
      new Attribute({
        key: 'dynamicHeight',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'headerPosition',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'selectedIndex',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(animationDone)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(focusChange)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(selectedIndexChange)',
        value: '',
        type: 'kv'
      })
    );
  }
  get template() {
    const template = `<mat-tab-group %style% %class% %bCustomProps% %animationDuration% %backgroundColor% %color% %dynamicHeight% %headerPosition% %selectedIndex% %(animationDone)% %(focusChange)% %(selectedIndexChange)%></mat-tab-group>`;
    return template;
  }
  set template(templateString) {}
};
