// @jatahworx/bhive-toolkits/angular-material/component/Toolbar.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class Toolbar extends Component {
  constructor() {
    const name = 'NG-Toolbar';
    const designerTemplate = `<div class=\"drop ad-flex\" component-label="Toolbar"></div>`;
    const paletteTemplate = 'Toolbar';
    const componentLabel = 'Toolbar';
    const templateUrl =
      'articles/#!components-guide-7/toolbar';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NAVIGATION.val);

    super.addAttribute(new Attribute({
      key: 'color',
      value: '',
      type: 'kv'
    }));
    super.addAttribute(new Attribute({
      key: 'content',
      value: '',
      type: 'a'
    }));
  }
  get template() {
    const template = `<mat-toolbar %bCustomProps% %color% %style% %class%>%content%</mat-toolbar>`;
    return template;
  }
  set template(templateString) {}
};
