// @jatahworx/bhive-toolkits/angular-material/component/VerticalBarChart.js
'use strict';
let Attribute = require('../../core/Attributes');
let Component = require('../../core/Component');

module.exports = class VerticalBarChart extends Component {
  constructor() {
    const name = 'NG-bar-chart-vertical';
    const designerTemplate = `<div class=\"ad-charts ad-vertical-bar-chart \" component-label="Vertical Bar Chart"></div>`;
    const paletteTemplate = 'Vertical Bar Chart';
    const componentLabel = 'Vertical Bar Chart';
    const templateUrl =
      'articles/#!components-guide-7/ngx-stacked-vertical-bar-chart';

    super({
      name: name,
      designerTemplate: designerTemplate,
      paletteTemplate: paletteTemplate,
      componentLabel: componentLabel,
      templateUrl: templateUrl
    }
    );
    super.setType(Component.COMPONENT_TYPE_TITLES.NGX_CHARTS.val);

    super.addAttribute(
      new Attribute({
        key: '[gradient]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxis]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[legend]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showXAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[showYAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[yAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[xAxisLabel]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '(select)',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[scheme]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: '[results]',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(
      new Attribute({
        key: 'fxLayout',
        value: '',
        type: 'kv'
      })
    );
    super.addAttribute(new Attribute({
      key: 'fxFlex',
      value: '',
      type: 'kv'
    }));
  }
  get template() {
    const template = `<div %fxFlex% %fxLayout%><ngx-charts-bar-vertical %bCustomProps% %[scheme]% %[results]% %style% %class% %[gradient]% %[xAxis]% %[yAxis]% %[legend]% %[showXAxisLabel]% %[showYAxisLabel]% %[yAxisLabel]% %[xAxisLabel]% %(select)%></ngx-charts-bar-vertical></div>`;
    return template;
  }
  set template(templateString) {}
};
