// @jatahworx/bhive-toolkits/angular-material/component/AdvanceInput/InputLabel.js
"use strict";
let AdvancedComponent = require("../../../core/AdvancedComponent");
let Attribute = require('../../../core/Attributes');

module.exports = class InputLabel extends AdvancedComponent {
  constructor() {
    const name = "ng-input-label";
    const designerTemplate = '<ng-input-label></ng-input-label>';
    const paletteTemplate = 'Input Label';
    const componentLabel = 'Input Label';

    super({
      name,
      designerTemplate,
      paletteTemplate,
      componentLabel,
      isAdvancedChild: true,
      templateUrl: "articles/#!components-guide-7/input"
    });
    super.addAttribute(new Attribute({
      key: 'Label',
      value: '',
      type: 'a',
      visible: false,
      isVisibleForParent: true
    }))
    super.composeTemplate();
  }

  get template() {
    let returnTemplate = ''; 
    let inputLableAttribute = this.getHtmlAttributes(this.htmlAttributes);
    if(inputLableAttribute['Label']._value.trim()!=""){
      returnTemplate = `<mat-label>%Label%</mat-label>`;
    }
    return returnTemplate;
  }

  set template(t) {}
};
