// @jatahworx/bhive-toolkits/angular-material/component/AdvanceInput/InputSuffix.js
"use strict";
let AdvancedComponent = require("../../../core/AdvancedComponent");
let Attribute = require('../../../core/Attributes');

module.exports = class InputSuffix extends AdvancedComponent {
  constructor() {
    const name = "ng-input-suffix";
    const designerTemplate = '<ng-input-suffix></ng-input-suffix>';
    const paletteTemplate = 'Input Suffix';
    const componentLabel = 'Input Suffix';

    super({
      name,
      designerTemplate,
      paletteTemplate,
      componentLabel,
      isAdvancedChild: true,
      templateUrl: "articles/#!components-guide-7/input"
    });
    super.addAttribute(new Attribute({
      key: 'Suffix',
      value: '',
      type: 'a',
      visible: false,
      isVisibleForParent: true
    }));
    super.composeTemplate();
  }

  get template() {
    let returnTemplate = '';
    let suffixAttribute = this.getHtmlAttributes(this.htmlAttributes);
    if(suffixAttribute['Suffix']._value.trim()!=""){
      returnTemplate = `<span matSuffix>%Suffix%</span>`;
    }
    return returnTemplate;
  }

  set template(t) {}
};
