// @jatahworx/bhive-toolkits/angular-material/component/AdvanceInput/InputPrefix.js
"use strict";
let AdvancedComponent = require("../../../core/AdvancedComponent");
let Attribute = require('../../../core/Attributes');

module.exports = class InputPrefix extends AdvancedComponent {
  constructor() {
    const name = "ng-input-prefix";
    const designerTemplate = '<ng-input-prefix></ng-input-prefix>';
    const paletteTemplate = 'Input Prefix';
    const componentLabel = 'Input Prefix';

    super({
      name,
      designerTemplate,
      paletteTemplate,
      componentLabel,
      isAdvancedChild: true,
      templateUrl: "articles/#!components-guide-7/input"
    });
    super.addAttribute(new Attribute({
      key: 'Prefix',
      value: '',
      type: 'a',
      visible: false,
      isVisibleForParent: true
    }));
    super.composeTemplate();
  }

  get template() {
    let returnTemplate = '';
    let prefixAttribute = this.getHtmlAttributes(this.htmlAttributes);
    if(prefixAttribute['Prefix']._value.trim()!=""){
      returnTemplate = `<span matPrefix>%Prefix%</span>`;
    }
    return returnTemplate;
  }

  set template(t) {}
};
