// @jatahworx/bhive-toolkits/angular-material/component/Html5/CustomHtmlEditor.js
'use strict';
const Attribute = require('../../../core/Attributes');
const AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CustomHtmlEditor extends AdvancedComponent {
  constructor() {
    const name = 'html-editor';
    const designerTemplate = ` <html-editor  b-props="html-editor" class=\"flex-row display-block non-dropable-comp-border\" component-label="HTML Editor" block-copy>
                                   </html-editor>`;
    const paletteTemplate = 'HTML Editor';
    const templateUrl = 'articles/#!components-guide-7/html';

    super({
      name,
      designerTemplate,
      paletteTemplate,
      isAdvancedChild: true,
      templateUrl,
    });

    super.addAttribute(new Attribute({
      key: 'html',
      value: '',
      type: 'a'
    }));

    super.composeTemplate({
      styles: `:host {
              padding: 1.5em;
              background: white;
              position: relative;
              margin: 10px 5px 3px 5px !important;
              border-radius: 5px !important;
              color: #404041;
              }`
    })
  }

  get template() {
    const template = '%html%';
    return template;
  }
  set template(templateString) {}

};
