// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardImage.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardImage extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-image';
        const designerTemplate = `<ng-card-image slot="cards" class=\" display-block ad-card-header flex-row\" block-copy>
        <div class=\"ad-card-avatar\"></div>
        <div class="flex-column flex-justify-center">
            <div class=\"ad-card-header-title ad-card-title-color ad-card-image-text\">Card Image</div>
        </ng-card-image>`;
        const paletteTemplate = 'Card Image';
        const componentLabel = 'Card Image';

        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            componentLabel,
            paletteTemplate,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(new Attribute({
            key: 'alt',
            value: '',
            type: 'kv'
        }));
        super.addAttribute(new Attribute({
            key: 'imgSrc',
            value: '',
            type: 'kv'
        }));
        super.addAttribute(new Attribute({
          key: '[src]',
          value: '',
          type: 'kv'
        }));
        super.addAttribute(
            new Attribute({
              key: 'imgSecure',
              value: '',
              type: 'a'
            })
          );
          super.addAttribute(
            new Attribute({
              key: '[collectionName]',
              value: '',
              type: 'kv'
            })
          );
          super.addAttribute(
            new Attribute({
              key: '[imageFilter]',
              value: '',
              type: 'kv'
            })
          );

        this.template = `
        <img mat-card-image %imgSecure% %[collectionName]% %[imageFilter]% %bCustomProps% %style% %class% %alt% %imgSrc% %[src]%>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 0.5em 1em;
              position: relative;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: black;
              }
              
              ::slotted(.card-image-text) {
                position: absolute;
                bottom: 0px;
                text-align: center;
                width: 90%;
                font-size: 14px;
                margin-bottom: 10px;
                color: #404041;
               }
               ::slotted(.ad-card-image) {
                   line-height: 40px;
               }
               `
        });
    }
};
