// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardTitle.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardTitle extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-title';
        const designerTemplate = `<ng-card-title slot="cards" block-copy class="display-block">
        <span class="component-placeholder title-align">Card Title</span>
      </ng-card-title>`;
        const paletteTemplate = 'Card Title';
        const componentLabel = 'Card Title';

        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            componentLabel,
            paletteTemplate,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(
            new Attribute({
              key: 'Title',
              value: 'Card Title',
              type: 'a',
              useAsLabel: true,
            })
          );
          super.addAttribute(new Attribute({
            key: 'align',
            value: '',
            type: 'kv',
          }));

        this.template = `
        <mat-card-title %class% %style% %align% %bCustomProps%>%Title%</mat-card-title>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 0.8em;
              position: relative;
              min-height: 1em;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: black;
              }`
          });
    }
};
