// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardSubTitle.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardSubTitle extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-subtitle';
        const designerTemplate = `<ng-card-subtitle slot="cards" block-copy class="display-block">
        <span class="component-placeholder title-align">Card Sub-Title</span>
      </ng-card-subtitle>`;
        const paletteTemplate = 'Card Sub-title';
        const componentLabel = 'Card Sub-title';

        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            componentLabel,
            paletteTemplate,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(
            new Attribute({
              key: 'Sub Title',
              value: 'Card Sub-Title',
              type: 'a',
              useAsLabel: true,
              isVisibleForParent:true
            })
          );
          super.addAttribute(new Attribute({
            key: 'align',
            value: '',
            type: 'kv',
          }));

        this.template = `
        <mat-card-subtitle %class% %style% %align% %bCustomProps%>%Sub Title%</mat-card-subtitle>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 1.3em;
              position: relative;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              }`
          });
    }
};
