// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardHeader.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardHeader extends AdvancedComponent {
    constructor() {
        const name = 'ng-cardheader';
        const designerTemplate = ` <ng-cardheader slot="cards" class=\"ad-card-header flex-row display-block\" block-copy>
                                        <div class=\"ad-card-avatar\"></div>
                                        <div class="flex-column flex-justify-center">
                                            <div class=\"ad-card-header-title\">Title</div>
                                            <div class=\"ad-card-subtitle\">Subtitle</div>
                                        </div>
                                   </ng-cardheader>`;
        const paletteTemplate = 'Card Header';
        const componentLabel = 'Card Header';

        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            componentLabel,
            paletteTemplate,
            isAdvancedChild: true,
            templateUrl
        });

        this.template = `
        <mat-card-header %bCustomProps% %style% %class%>
            <mat-card-title>%title%</mat-card-title>
            <mat-card-subtitle>%subtitle%</mat-card-subtitle>
            <img mat-card-avatar %imgSecure% %[collectionName]% %[imageFilter]% %imgSrc% %[src]% %alt%>
        </mat-card-header>
        `

        super.addAttribute(new Attribute({
            key: 'title',
            value: 'title',
            type: 'a'
        }));
        super.addAttribute(
            new Attribute({
                key: 'subtitle',
                value: 'subtitle',
                type: 'a'
            })
        );
        super.addAttribute(new Attribute({
            key: 'imgSrc',
            value: '',
            type: 'kv'
        }));
        super.addAttribute(new Attribute({
            key: '[src]',
            value: '',
            type: 'kv'
          }));
        super.addAttribute(new Attribute({
            key: 'alt',
            value: '',
            type: 'kv'
        }));
        super.addAttribute(
            new Attribute({
              key: 'imgSecure',
              value: '',
              type: 'a'
            })
          );
          super.addAttribute(
            new Attribute({
              key: '[collectionName]',
              value: '',
              type: 'kv'
            })
          );
          super.addAttribute(
            new Attribute({
              key: '[imageFilter]',
              value: '',
              type: 'kv'
            })
          );
        super.composeTemplate({
            styles: `:host {
              padding: 0.7em;
              background: white;
              position: relative;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: #404041;
              }`
          })
    }
};
