// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardAction.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardAction extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-action';
        const designerTemplate = ` <ng-card-action slot="cards" class=\"drop display-block\" component-label="Card Action" block-copy>
        </ng-card-action>`;
        const paletteTemplate = 'Card Action';
        const componentLabel = 'Card Action';
        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            paletteTemplate,
            componentLabel,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(
            new Attribute({
              key: 'align',
              value: '',
              type: 'kv'
            })
          );

        this.template = `
        <mat-card-actions %align% %bCustomProps% %style% %class%></mat-card-actions>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 0.7em !important;
              position: relative;
              min-height: 1em;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: black;
              }`
          });
    }
};
