// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardFooter.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardFooter extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-footer';
        const designerTemplate = `<ng-card-footer class=\"drop display-block\" slot="cards" component-label="Card Footer" block-copy></ng-card-footer>`;
        const paletteTemplate = 'Card Footer';
        const componentLabel = 'Card Footer';
        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            paletteTemplate,
            componentLabel,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(
            new Attribute({
              key: 'align',
              value: '',
              type: 'kv'
            })
          );

        this.template = `
        <mat-card-footer %align% %bCustomProps% %style% %class%></mat-card-footer>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 0.7em !important;
              position: relative;
              min-height: 1em;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: black;
              }`
          });
    }
};
