// @jatahworx/bhive-toolkits/angular-material/component/AdvanceCard/CardContent.js
'use strict';
let Attribute = require('../../../core/Attributes');
let AdvancedComponent = require("../../../core/AdvancedComponent");

module.exports = class CardContent extends AdvancedComponent {
    constructor() {
        const name = 'ng-card-content';
        const designerTemplate = ` <ng-card-content slot="cards" class=\"drop display-block\" component-label="Card Content" block-copy>
        </ng-card-content>`;
        const paletteTemplate = 'Card Content';
        const componentLabel = 'Card Content';        
        const templateUrl = 'articles/#!components-guide-7/card';

        super({
            name,
            template: ``,
            designerTemplate,
            paletteTemplate,
            componentLabel,
            isAdvancedChild: true,
            templateUrl
        });

        super.addAttribute(
            new Attribute({
              key: 'align',
              value: '',
              type: 'kv'
            })
          );

        this.template = `
        <mat-card-content %align% %style% %class% %bCustomProps%></mat-card-content>
        `;
        super.composeTemplate({
            styles: `:host {
              padding: 0.7em !important;
              position: relative;
              min-height: 1em;
              border: 1px solid lightgrey;
              margin: 6px 25px !important;
              border-radius: 5px !important;
              color: black;
              }
              `
          });
    }
};
