// @jatahworx/bhive-toolkits/ServiceDesigner/ndefault-pub-sub/nodes/server/emit.js
let BaseComponent = require('@jatahworx/bhive-toolkits').BaseComponent;

module.exports = class emit extends BaseComponent {
	constructor(constructorOptions) {
		super(
			constructorOptions.id,
			constructorOptions.name,
			'neutrinos_service_designer_' + 'emit',

			'emit'
		);
        this.constructorOptions = constructorOptions;
	}
	/**
	 * returns Array of {
	 *  libraryName: string,
	 *  modules: [string],
	 *  alias: string
	 * }
	 */
	generateImports(rootPaths) {
		return [
			{
				library: rootPaths.pathToUtilsRoot + '/ndefault-pub-sub/listen/listen',
				modules: [
					"pubsubUtil"
				]
			}
		];
	}
	/**
	 * @typedef {Object} ServiceVariable
	 * @property {string} name - Name of the variable
	 * @property {string} [type] - Type of the variable
	 * @property {any} [value] - Initial value to assign to the variable
	 */

	/**
	 * @returns {Array<ServiceVariable>}
	 */
	declareVariables() {
		console.log(
			this.name,
			'Base Class declare variable called.. returning empty array'
		);
		return [];
	}

	generateSnippet(serviceType, serviceClassTemplate) {
		if (!this.constructorOptions.eventName ) {
            throw new Error(`Emit Node: Event name can't be empty`);
        }
		return `let pubsubInstance = pubsubUtil.getInstance();
        pubsubInstance.emit(${this.constructorOptions.eventName}, bh)`;
	}

	getCallTemplate() {
		console.log(
			this.name,
			'Base Class generation called.. returning empty call template..'
		);
		return '';
	}

	populateConfig(configObj) {}

	/**
	 * reutrn [{name: '', version: ''}]
	 */
	updatePackageJSON() {
		return [];
	}

	updateTemplate(serviceClassTemplate) {
		// base calss update Template. return as is;
		return serviceClassTemplate;
	}

	getErrorTemplate() {
		return `
        export class errorCheck {
                //appendnew_node
        }
        `;
	}
};
